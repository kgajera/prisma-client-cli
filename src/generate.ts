import type {
  DMMF,
  EnvValue,
  GeneratorOptions,
} from "@prisma/generator-helper";
import { parseEnvValue } from "@prisma/internals";
import { ModuleKind, Project, ScriptTarget } from "ts-morph";
import { generateBin } from "./generators/bin";
import { generateOperationCommand } from "./generators/operationCommand";
import { generateQueryCommand } from "./generators/queryCommand";
import { generateTopicCommand } from "./generators/topicCommand";
import { isDisabled } from "./utils/isDisabled";

const UNSUPPORTED_OPERATIONS = [
  "aggregate",
  "createMany",
  "findFirstOrThrow",
  "findUniqueOrThrow",
  "groupBy",
];

export async function generate({ dmmf, generator }: GeneratorOptions) {
  if (isDisabled()) {
    return;
  }

  const outputDir = parseEnvValue(generator.output as EnvValue);

  const project = new Project({
    compilerOptions: {
      outDir: outputDir,
      module: ModuleKind.NodeNext,
      target: ScriptTarget.ES2020,
      declaration: false,
    },
  });

  project.createSourceFile("index.ts", "export {run} from '@oclif/core'", {
    overwrite: true,
  });

  generateBin(outputDir);
  generateQueryCommand(project);

  const outputTypes: DMMF.SchemaField[] = [];
  for (const t of dmmf.schema.outputObjectTypes.prisma) {
    outputTypes.push(...t.fields);
  }

  for (const model of dmmf.datamodel.models) {
    generateTopicCommand(project, model);

    const modelOperations = dmmf.mappings.modelOperations.find(
      (mo) => mo.model === model.name
    );

    if (modelOperations) {
      const { model: modelName, ...operations } = modelOperations;

      for (const operationName in operations) {
        if (UNSUPPORTED_OPERATIONS.includes(operationName)) {
          continue;
        }

        const operationArgs = outputTypes.find(
          (f) => f.name === operations[operationName as keyof typeof operations]
        );
        generateOperationCommand(
          project,
          model,
          operationName,
          operationArgs,
          dmmf
        );
      }
    }
  }

  await project.emit();
}
