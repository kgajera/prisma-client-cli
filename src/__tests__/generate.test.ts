import fs from "fs";
import os from "os";
import path from "path";
import { expect, test } from "vitest";
import { getSampleDMMF } from "./__fixtures__/getSampleDMMF";
import { generate } from "../generate";

test("generate", async () => {
  const dmmf = await getSampleDMMF();
  const tmpOutputDir = path.join(os.tmpdir(), "prisma-cli-generator-test");

  await generate({
    dmmf,
    generator: {
      output: {
        fromEnvVar: null,
        value: tmpOutputDir,
      },
    },
  } as any);

  [
    "QueryCommand.js",
    path.join("bin", "run"),
    path.join("commands", "user", "index.js"),
    path.join("commands", "user", "findMany.js"),
    path.join("commands", "user", "deleteOne.js"),
  ].forEach((file) => {
    const content = fs.readFileSync(path.resolve(tmpOutputDir, file), "utf-8");
    expect(content).toMatchSnapshot();
  });

  fs.rmSync(tmpOutputDir, { recursive: true, force: true });
}, 10000);
