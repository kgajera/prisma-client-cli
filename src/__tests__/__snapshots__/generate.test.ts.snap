// Vitest Snapshot v1

exports[`generate 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.QueryCommand = void 0;
const core_1 = require(\\"@oclif/core\\");
const client_1 = require(\\"@prisma/client\\");
class QueryCommand extends core_1.Command {
    async init() {
        await super.init();
        const { flags } = await this.parse(this.constructor);
        this.flags = flags;
    }
    getQueryArgs() {
        return Object.keys(this.flags).reduce((acc, path) => {
            const keys = path.split('.');
            let cur = acc;
            if (path === \\"_csv\\") {
                return acc;
            }
            keys.forEach((key, i) => {
                if (i < keys.length - 1) {
                    cur[key] = cur[key] || {};
                    cur = cur[key];
                }
                else {
                    cur[key] = this.flags[path];
                }
            });
            return acc;
        }, {});
    }
    async outputQueryResult(data) {
        const isArray = Array.isArray(data);
        if (data === null) {
            console.log(\\"Record not found\\");
            return;
        }
        else if (isArray && data.length === 0) {
            console.log(\\"No records found\\");
            return;
        }
        const dataArray = isArray ? data : [data], keys = Object.keys(dataArray[0]), columns = {};
        for (const key of keys) {
            columns[key] = {};
        }
        core_1.CliUx.ux.table(dataArray, columns, { csv: this.flags[\\"_csv\\"] ?? false });
    }
    async run() {
        const client = new client_1.PrismaClient(), args = this.getQueryArgs(), result = await this.query(client, args);
        this.outputQueryResult(result);
    }
}
exports.QueryCommand = QueryCommand;
QueryCommand.globalFlags = {
    _csv: core_1.Flags.boolean(),
};
"
`;

exports[`generate 2`] = `
"#!/usr/bin/env node

const oclif = require('@oclif/core')

oclif.run().then(require('@oclif/core/flush')).catch(require('@oclif/core/handle'))"
`;

exports[`generate 3`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const core_1 = require(\\"@oclif/core\\");
class UserCommand extends core_1.Command {
    async run() {
        const help = new core_1.Help(this.config);
        help.showTopicHelp({ name: 'user' });
    }
}
exports.default = UserCommand;
UserCommand.hidden = true;
"
`;

exports[`generate 4`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const core_1 = require(\\"@oclif/core\\");
const QueryCommand_1 = require(\\"../../QueryCommand\\");
class UserFindMany extends QueryCommand_1.QueryCommand {
    async query(client, args) {
        return client.user.findMany(args);
    }
}
exports.default = UserFindMany;
UserFindMany.flags = {
    ['where.id.equals']: core_1.Flags.integer({}),
    ['where.id.in']: core_1.Flags.integer({}),
    ['where.id.notIn']: core_1.Flags.integer({}),
    ['where.id.lt']: core_1.Flags.integer({}),
    ['where.id.lte']: core_1.Flags.integer({}),
    ['where.id.gt']: core_1.Flags.integer({}),
    ['where.id.gte']: core_1.Flags.integer({}),
    ['where.id.not']: core_1.Flags.integer({}),
    ['where.id']: core_1.Flags.integer({}),
    ['where.email.equals']: core_1.Flags.string({}),
    ['where.email.in']: core_1.Flags.string({}),
    ['where.email.notIn']: core_1.Flags.string({}),
    ['where.email.lt']: core_1.Flags.string({}),
    ['where.email.lte']: core_1.Flags.string({}),
    ['where.email.gt']: core_1.Flags.string({}),
    ['where.email.gte']: core_1.Flags.string({}),
    ['where.email.contains']: core_1.Flags.string({}),
    ['where.email.startsWith']: core_1.Flags.string({}),
    ['where.email.endsWith']: core_1.Flags.string({}),
    ['where.email.not']: core_1.Flags.string({}),
    ['where.email']: core_1.Flags.string({}),
    ['where.name.equals']: core_1.Flags.string({}),
    ['where.name.in']: core_1.Flags.string({}),
    ['where.name.notIn']: core_1.Flags.string({}),
    ['where.name.lt']: core_1.Flags.string({}),
    ['where.name.lte']: core_1.Flags.string({}),
    ['where.name.gt']: core_1.Flags.string({}),
    ['where.name.gte']: core_1.Flags.string({}),
    ['where.name.contains']: core_1.Flags.string({}),
    ['where.name.startsWith']: core_1.Flags.string({}),
    ['where.name.endsWith']: core_1.Flags.string({}),
    ['where.name.not']: core_1.Flags.string({}),
    ['where.name']: core_1.Flags.string({}),
    ['where.profile.id']: core_1.Flags.integer({}),
    ['where.profile.bio']: core_1.Flags.string({}),
    ['where.profile.userId']: core_1.Flags.integer({}),
    ['cursor.id']: core_1.Flags.integer({}),
    ['cursor.email']: core_1.Flags.string({}),
    ['take']: core_1.Flags.integer({}),
    ['skip']: core_1.Flags.integer({}),
};
"
`;

exports[`generate 5`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const core_1 = require(\\"@oclif/core\\");
const QueryCommand_1 = require(\\"../../QueryCommand\\");
class UserDeleteOne extends QueryCommand_1.QueryCommand {
    async query(client, args) {
        return client.user.delete(args);
    }
}
exports.default = UserDeleteOne;
UserDeleteOne.flags = {
    ['where.id']: core_1.Flags.integer({}),
    ['where.email']: core_1.Flags.string({}),
};
"
`;
